documentation/public/internals/design-docs/language-definition-v2/index.md

```diff
@@ -0,0 +1,438 @@
+                fields = (operator = Required(Terminal([Plus])))
+            )]
+        ),
+        PrecedenceExpression(
+            name = FunctionCallExpression,
+            operators = [PrecedenceOperator(
+                model = Postfix,
+                fields = (
+                    open_paren = Required(Terminal([OpenParen])),
+                    arguments = Required(NonTerminal(Arguments)),
+                    close_paren = Required(Terminal([CloseParen]))
+                )
+            )]
+        ),
+        PrecedenceExpression(
+            name = NegationExpression,
+            operators = [PrecedenceOperator(
+                model = Prefix,
+                fields = (operator = Required(Terminal([Not])))
+            )]
+        )
+    )],
+    default_primary_expression = Identifier,
+    primary_expressions = [
+        PrimaryExpression(expression = Identifier),
+        PrimaryExpression(expression = NumberLiteral),
+        PrimaryExpression(expression = StringLiteral)
+    ]
+)
+```
+
+```rust title="AST Type"
+pub enum Expression {
+    AdditionExpression { expression: Rc<AdditionExpression> },
+    FunctionCallExpression { expression: Rc<FunctionCallExpression> },
+    NegationExpression { expression: Rc<NegationExpression> },
+
+    Identifier { expression: Rc<TerminalNode> },
+    NumberLiteral { expression: Rc<TerminalNode> },
+    StringLiteral { expression: Rc<TerminalNode> },
+}
+
+pub struct AdditionExpression {
+    // 'left_operand' auto-generated (before) because it is a binary expression, and same type as parent
+    pub left_operand: Rc<Expression>,
+    // operator 'fields' are flattened into the expression node here
+    pub operator: Rc<TerminalNode>,
+    // 'right_operand' auto-generated (after) because it is a binary expression, and same type as parent
+    pub right_operand: Rc<Expression>,
+
+    pub cst: Rc<NonTerminalNode>,
+}
+
+pub struct FunctionCallExpression {
+    // 'operand' auto-generated (before) because it is a postfix expression, and same type as parent
+    pub operand: Rc<Expression>,
+    // operator 'fields' are flattened into the expression node here
+    pub open_paren: Rc<TerminalNode>,
+    pub arguments: Rc<Arguments>,
+    pub close_paren: Rc<TerminalNode>,
+
+    pub cst: Rc<NonTerminalNode>,
+}
+
+pub struct NegationExpression {
+    // operator 'fields' are flattened into the expression node here
+    pub operator: Rc<TerminalNode>,
+    // 'operand' auto-generated (after) because it is a prefix expression, and same type as parent
+    pub operand: Rc<Expression>,
+
+    pub cst: Rc<NonTerminalNode>,
+}
+```
+
+## Error Recovery
+
+This is a newly introduced feature which I'm less familiar with, so I'm looking for validation on whether the ideas below
+are possible (in a reasonable time), or if we should look for a better solution:
+
+I believe it is important to provide the AST, even if there are errors in the input. This is because we want to avoid forcing users
```

Comment:
I think this is not important, because analyses are typically either possible or not on partial inputs. So I am 100% happy with valid-only AST production. In particular I think we should at this time not pay for that feature in either time or complexity.

Reply:
Removed.